crates/swc_ecma_lints/src/rules/no_alert.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{
+    config::{LintRuleReaction, RuleConfig},
+    rule::{visitor_rule, Rule},
+};
+use swc_atoms::JsWord;
+use swc_common::{collections::AHashSet, errors::HANDLER, Span, SyntaxContext};
+use swc_ecma_ast::*;
+use swc_ecma_utils::{collect_decls_with_ctxt, ident::IdentLike};
+use swc_ecma_visit::{noop_visit_type, Visit, VisitWith};
+
+const FN_NAMES: &[&str] = &["alert", "confirm", "prompt"];
+const GLOBAL_THIS_PROP: &str = "globalThis";
+const OBJ_NAMES: &[&str] = &["window", GLOBAL_THIS_PROP];
+
+pub fn no_alert(
+    program: &Program,
+    config: &RuleConfig<()>,
+    top_level_ctxt: SyntaxContext,
+    es_version: EsVersion,
+) -> Option<Box<dyn Rule>> {
+    let top_level_declared_vars: AHashSet<Id> = collect_decls_with_ctxt(program, top_level_ctxt);
+    let rule_reaction = config.get_rule_reaction();
+
+    match rule_reaction {
+        LintRuleReaction::Off => None,
+        _ => Some(visitor_rule(NoAlert::new(
+            *rule_reaction,
+            top_level_declared_vars,
+            top_level_ctxt,
+            es_version,
+        ))),
+    }
+}
+
+#[derive(Debug, Default)]
+struct NoAlert {
+    expected_reaction: LintRuleReaction,
+    top_level_ctxt: SyntaxContext,
+    top_level_declared_vars: AHashSet<Id>,
+    pass_call_on_global_this: bool,
+}
+
+impl NoAlert {
+    fn new(
+        expected_reaction: LintRuleReaction,
+        top_level_declared_vars: AHashSet<Id>,
+        top_level_ctxt: SyntaxContext,
+        es_version: EsVersion,
+    ) -> Self {
+        Self {
+            expected_reaction,
+            top_level_ctxt,
+            top_level_declared_vars,
+            pass_call_on_global_this: es_version < EsVersion::Es2020,
+        }
+    }
+
+    fn emit_report(&self, span: Span, fn_name: &str) {
+        let message = format!("Unexpected {}", fn_name);
+
+        HANDLER.with(|handler| match self.expected_reaction {
+            LintRuleReaction::Error => {
+                handler.struct_span_err(span, &message).emit();
+            }
+            LintRuleReaction::Warning => {
+                handler.struct_span_warn(span, &message).emit();
+            }
+            _ => {}
+        });
+    }
+
+    fn check(&self, call_span: Span, obj: &Option<JsWord>, prop: &JsWord) {
+        if let Some(obj) = obj {
+            let obj_name: &str = &*obj;
+
+            if self.pass_call_on_global_this && obj_name == GLOBAL_THIS_PROP {
+                return;
+            }
+
+            if !OBJ_NAMES.contains(&obj_name) {
+                return;
+            }
+        }
+
+        let fn_name: &str = &*prop;
+
+        if FN_NAMES.contains(&fn_name) {
+            self.emit_report(call_span, fn_name);
+        }
+    }
+}
+
+impl Visit for NoAlert {
+    noop_visit_type!();
+
+    fn visit_call_expr(&mut self, call_expr: &CallExpr) {
+        let mut call_collector = CallCollector {
```

Comment:
Using a nested visitor without a stopper method is bad for performance.

Reply:
Also, we don't need two visitor here, and visitor is quite big in final binary.

Reply:
(That's why methods for collecting bindings uses single visitor)

Reply:
Hmm.. Seems like I need advice how visit indents/members only on callee

Reply:
I rewrote implementation without nested visitor, but it looks a bit awkward by my mind